metagraph/wrappers.py

```diff
@@ -148,3 +149,79 @@ def assert_equal(
 
 class BipartiteGraphWrapper(Wrapper, abstract=BipartiteGraph, register=False):
     pass
+
+
+class NodeEmbeddingWrapper(Wrapper, abstract=NodeEmbedding, register=False):
+    def __init__(self, matrix, nodes=None):
+        super().__init__()
+        self._assert(
+            type(matrix).Type.compute_abstract_properties(matrix, {"is_dense"})[
+                "is_dense"
+            ],
+            f"Matrix {matrix} must be dense",
+        )
+        self.matrix = matrix
+        if nodes is not None:
+            nodes_dtype = type(nodes).Type.compute_abstract_properties(
+                nodes, {"dtype"}
+            )["dtype"]
+            self._assert(
+                nodes_dtype == "int",
+                f"Node map {nodes} must have dtype of int rather than {nodes_dtype}",
+            )
+        self.nodes = nodes
+
+    class TypeMixin:
+        @classmethod
+        def _compute_abstract_properties(
+            cls, obj, props: Set[str], known_props: Dict[str, Any]
+        ) -> Dict[str, Any]:
+            ret = known_props.copy()
+
+            # fast properties
+            for prop in {"matrix_dtype"} - ret.keys():
+                if prop == "matrix_dtype":
+                    ret[prop] = type(obj.matrix).Type.compute_abstract_properties(
+                        obj.matrix, {"dtype"}, {}
+                    )
+
+            return ret
+
+        @classmethod
+        def assert_equal(
+            cls,
+            obj1,
+            obj2,
+            aprops1,
+            aprops2,
+            cprops1,
+            cprops2,
+            *,
+            rel_tol=1e-9,
+            abs_tol=0.0,
+        ):
+            assert aprops1 == aprops2, f"property mismatch: {aprops1} != {aprops2}"
+
+            matrix_class = type(obj1.matrix).Type
+            matrix_class.assert_equal(
+                obj1.matrix,
+                obj2.matrix,
+                {"dtype": aprops1["dtype"]},
+                {"dtype": aprops2["dtype"]},
+                {},
+                {},
+                rel_tol=rel_tol,
+                abs_tol=abs_tol,
+            )
+
+            nodes_class = type(obj1.nodes).Type
+            nodes_class.assert_equal(
+                obj1.nodes,
+                obj2.nodes,
+                {"dtype": aprops1["dtype"]},
```

Comment:
The node map's dtype is always "int". No need to reference `aprops` here.

Reply:
https://github.com/metagraph-dev/metagraph/pull/133/commits/2834fe217a903c7a1196d2990aaab3ac94e6c515 removes this reference to `aprops`.